### ISSUE TITLE
ImageBuffer.dispose() causes crash

### ISSUE THREAD

**Issue by zathras** [flutter/flutter#83764]
Code is like this:
```
    final buf = await ui.ImmutableBuffer.fromUint8List(encoded);
    final des = _descriptor = await ui.ImageDescriptor.encoded(buf);
    final codec = _codec = await des.instantiateCodec();


...  Later, when we're done with the images ...
    try {
        des.dispose();
    } catch (e) {
        print('https://github.com/flutter/flutter/issues/83421 $e');
    }
    codec.dispose();
    buf.dispose();   // <-- eliminating this makes memory corruption go away
```
It's not documented whether `ImageDescriptor.encoded(ImmutableBuffer)` takes over ownership of the ImageBuffer (and calls dispose() itself), or if the caller is responsible for this.  It's also hard to tell what the intent is by experimentation, since this whole area is in flux, what with ImageDescriptor.dispose() not being implemented in the current release.

So, if the intent is that encoded() **does** take over ownership, this needs to be documented, which is one kind of bug.

If the intent is that encoded **does not** take over ownership, then the crash is a bug.

Further, in debug mode, ImageBuffer.dispose() on a disposed buffer should emit some kind of warning, which is, in fairness, a different kind of bug.  So, either way, there are two bugs (since an under-specified API is a bug, and a rather serious one, at that).
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.2.1, on macOS 11.2.3 20D91 darwin-arm, locale
    en-US)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] IntelliJ IDEA Community Edition (version 2021.1)
[✓] Connected device (2 available)

! Doctor found issues in 1 category.
```

**Comment by TahaTesser** [comment]
Hi @zathras 
Can you please provide a complete reproducible minimal code sample, and complete `flutter run --verbose` logs running the minimal code sample?
Thank you

**Comment by zathras** [comment]
In order for it to be possible to do that sensibly, the specification ambiguity needs to be resolved.  What is the intended specification?

**Comment by jason-simmons** [comment]
Is this still crashing on the latest master branch version of Flutter?

In particular, https://github.com/flutter/engine/commit/01c7bf739be20904b7dac7dda48ed5b90e6d1c2b landed a few days ago and includes improvements to these APIs.

In the tip-of-tree engine, both the `ImmutableBuffer` and the `ImageDescriptor` hold references to the raw data buffer.  Each of those objects will drop its reference when `dispose` is called or when the object is GCed.  The data buffer will be freed when the last reference is dropped.

@dnfield

**Comment by dnfield** [comment]
These should all be treated as shared pointers.

There were some bugs in the implementation of `dispose` on some of these on the C++ side with use after free and missing glue code. As @jason-simmons said you should be good with your code on master now.

**Comment by dnfield** [comment]
The crash was the bug, and should be fixed now because we don't have a use after free error anymore. You can see a full test of this here: https://github.com/flutter/engine/commit/01c7bf739be20904b7dac7dda48ed5b90e6d1c2b#diff-3235df72cf81469708c357a61a6835337560f2e179822bed60b3b648c530ef20R12-R28

Closing as fixed, please feel free to ping this issue if there's something I'm missing here :)

**Comment by zathras** [comment]
For the unresolved bugs,

See https://github.com/flutter/flutter/issues/83908, which I opened as a new bug because I don't have permission to reopen a bug.  This should be re-opened, since it's not resolved.

See also https://github.com/flutter/flutter/issues/83910, which isn't quite a bug, but it's close, and easy to fix.

**Comment by dnfield** [comment]
@zathras - Do you have something in mind for what's left to do here? It's definitely good to be more explicit about ownership in docs, but I'm curious about which doc you were expecting to find but did not.

Typically, if not specified, Dart objects should be thought of as regular GC-able objects. That is, if there is any path from that object to a GC root, the object is alive; otherwise it is not.

Sometimes we specify things more tightly to try to encourage people to do things like call dispose at the appropriate time.

I'm fine with further specifying the docs here, and would be happy to work on that with your or review a PR if you're interested in writing one.

**Comment by zathras** [comment]
Well, since these object have dispose() methods, they need extra care, precisely because having a dispose() that frees native resources means they can't be thought of as regular GC-able objects.  That's what finalizers (including in all their weak-references glory, around JDK 1.3 if memory serves) were trying to address, but of course that leads to a different problem (small objects holding big native resources aren't collected fast enough).  Even with finalizers, the analogous java.awt.Image method tripped up two versions of Java before it got fixed, and then the bug was re-introduced into Android, and then fixed there maybe a decade later.  I think this sort of thing is one of the motivations for Flutters stateful widget/managed lifetime model.

Anyway, the needed rule is that whenever there's a dispose(), users of the API need to be told under what circumstances they're responsible for calling it, and under what circumstances they absolutely must not.  It's not mere documentation, it's normative specification text that's essential to correctness.

Since there isn't unambiguous normative text, I kind of have to make some inferences to figure out the intended specification.  It sounds like the intent here is that the programmer using the API should never dispose() anything until everything is no longer being used?  That's surprising -- usually for these kinds of things, if there's a native object potentially shared by more than one dispose()able object, one does reference-counting.  That makes the API much more robust, in that it upholds the promise that programmers at the Dart level can never corrupt memory.  Plus, as Tim Lindholm likes to say, it follows the principle of least astonishment.

That also makes it possible to write, say, a function that takes a Uint8Array as input, and delivers an Image object.  The implementation of that function would create the ImageBuffer and codec internally, and then dispose() those two objects, secure in the knowledge that any shared native objects won't be freed until the returned Image is dispose()d.  Without the reference counting, you can't possibly write a correct, non-leaking version of that function -- you're required to call dispose(), but you can't.

So:  Is the intended behavior **really** to force users of the API to hang on to the intermediate products (the buffer and the codec), just so they can call dispose() on these intermediate objects when the image(s) is/are disposed?  If that really is the intended behavior, then one way to specify this would be text like the following on ImageBuffer.dispose() (and similar text on the codec):

Users of ImageBuffer shall call dispose() when the data in this buffer is not longer needed.  dispose() shall not be called until after objects that have been given a reference to this object are still being used.

Some informative text giving the example of a codec using the buffer would be helpful, too.

Google specification documentation tends to not like formal terms like "shall," but a more chatty style can be more prone to being ambiguous.

**Comment by dnfield** [comment]
We tend to use `must` in documentation to just make it clear.

I completely agree with your assessment, and am actively working on cleaning up some of the usage around this. Part of our problem right now is that some of these disposable objects have long been used without calling dispose at the right time(s) (or at all). 

My question was more about where you looked for documentation though - on the class itself? On the dispose method? Somewhere else?

**Comment by dnfield** [comment]
(And, off topic - this kind of issue is why I don't have sufficient time to work on the out of order defs stuff on flutter_svg at the moment!)

**Comment by zathras** [comment]
Oh, well, where I actually looked was pretty much "all over the place" - and the results were pretty sparse :-). Then when I saw the exception on the missing native method, I stopped looking.

But as to "where should it be specified," under normal circumstances (e.g. with a referenced-counted implementation), a description on dispose() is where I'd expect the normative part to be.  Some informative text spelling out the likely consequences of not calling dispose() can be helpful (and can include some dire language about running out of native heap).  The main class documentation should have an informative note referencing dispose.

In the case where there's sharing going on without reference counting to keep everything safe, and thus prohibitions against calling dispose() "too early," a normative specification that ownership is **not** being passed and/or tracked at the point of sharing would be needed.  A dire warning of the consequences of calling dispose() "too early" at this point would be helpful informative text to add.  So, in this case, that would be on (at least) ImageDescriptor.encoded() and ImageDescriptor.instantiateCodec().

In the case where the sharing is protected by e.g. reference counting, the text at the point of sharing can arguably be considered informative, but it's still helpful.  https://api.flutter.dev/flutter/dart-ui/Codec/getNextFrame.html gets it about right.  (BTW, that's an example of what I meant by the "somewhat chatty" style that's part of the Google culture - in a more formal context, it would say "the caller ... shall ensure that dispose is called" or "the caller ... ensure that dispose is called", rather than the more passive "is responsible for disposing."  Having more formal style guides, and editors to enforce them can feel like a bit of a straight jacket, but it also catches bugs before they happen.)
